### PR TITLE
Move the preallocate tests into their own files

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

In #1920, we put the preallocate tests in the same files as the TrustedPreallocate impls.

But in Zebra, we try to put tests in their own files.

## Solution

- move the preallocate tests into their own files
- move the MetaAddr proptest into its own file
- do some minor formatting and cleanups

## Review

Anyone can review. This code movement isn't urgent.

@preston-evans98 you might also want to review this, but don't worry if you're busy.

## Related Issues

Cleanup after #1920.

## Follow Up Work

Turn the InventoryHash and MetaAddr tests into proptests.

### CODE REVIEW

File: zebra-network/src/protocol/external/tests/preallocate.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    let serializd_inv_err = inv_err
+        .zcash_serialize_to_vec()
+        .expect("Serialization to vec must succeed");
+    assert!(serializd_inv_err.len() == INV_HASH_SIZE)
+}
+#[test]
+/// Verifies that...
+/// 1. The smallest disallowed vector of `InventoryHash`s is too large to fit in a legal Zcash message
+/// 2. The largest allowed vector is small enough to fit in a legal Zcash message
+fn meta_addr_max_allocation_is_correct() {
```
Comment (teor2345) [inline zebra-network/src/protocol/external/tests/preallocate.rs:47]: Typo + formatting
```suggestion

/// Verifies that...
/// 1. The smallest disallowed vector of `InventoryHash`s is too large to fit in a legal Zcash message
/// 2. The largest allowed vector is small enough to fit in a legal Zcash message
#[test]
fn inv_hash_max_allocation_is_correct() {
```

File: zebra-chain/src/sprout/tests/preallocate.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    /// This verifies that our calculated `TrustedPreallocate::max_allocation()` is indeed an upper bound.
+    fn joinsplit_btcv14_size_is_correct(joinsplit in <JoinSplit<Bctv14Proof>>::arbitrary_with(())) {
+        let serialized = joinsplit.zcash_serialize_to_vec().expect("Serialization to vec must succeed");
+        prop_assert!(serialized.len() as u64 == BCTV14_JOINSPLIT_SIZE)
+    }
+
+    #[test]
+    /// Confirm that each  JoinSplit<Btcv14Proof> takes exactly  GROTH16_JOINSPLIT_SIZE bytes when serialized.
+    /// This verifies that our calculated `TrustedPreallocate::max_allocation()` is indeed an upper bound.
+    fn joinsplit_groth16_size_is_correct(joinsplit in <JoinSplit<Groth16Proof>>::arbitrary_with(())) {
```
Comment (teor2345) [inline zebra-chain/src/sprout/tests/preallocate.rs]:
```suggestion
proptest! {
    /// Confirm that each  JoinSplit<Btcv14Proof> takes exactly BCTV14_JOINSPLIT_SIZE bytes when serialized.
    /// This verifies that our calculated `TrustedPreallocate::max_allocation()` is indeed an upper bound.
    #[test]
    fn joinsplit_btcv14_size_is_correct(joinsplit in <JoinSplit<Bctv14Proof>>::arbitrary_with(())) {
        let serialized = joinsplit.zcash_serialize_to_vec().expect("Serialization to vec must succeed");
        prop_assert!(serialized.len() as u64 == BCTV14_JOINSPLIT_SIZE)
    }

    /// Confirm that each  JoinSplit<Btcv14Proof> takes exactly  GROTH16_JOINSPLIT_SIZE bytes when serialized.
    /// This verifies that our calculated `TrustedPreallocate::max_allocation()` is indeed an upper bound.
    #[test]
    fn joinsplit_groth16_size_is_correct(joinsplit in <JoinSplit<Groth16Proof>>::arbitrary_with(())) {
```

File: zebra-chain/src/transaction/tests/preallocate.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+        // Check that our smallest_disallowed_vec is only one item larger than the limit
+        prop_assert!(((smallest_disallowed_vec.len() - 1) as u64) == <Arc<Transaction>>::max_allocation());
+        // Check that our smallest_disallowed_vec is too big to be included in a valid block
+        prop_assert!(serialized.len() as u64 > MAX_BLOCK_BYTES);
+    }
+
+    #[test]
+    /// Verify the smallest disallowed vector of `Input`s is too large to fit in a Zcash block
+    fn input_max_allocation_is_big_enough(input in Input::arbitrary()) {
```
Comment (teor2345) [inline zebra-chain/src/transaction/tests/preallocate.rs]:
```suggestion
    /// Verify the smallest disallowed vector of `Transaction`s is too large to fit in a Zcash block
    #[test]
    fn tx_max_allocation_is_big_enough(tx in Transaction::arbitrary()) {

        let max_allocation: usize = <Arc<Transaction>>::max_allocation().try_into().unwrap();
        let mut smallest_disallowed_vec = Vec::with_capacity(max_allocation + 1);
        for _ in 0..(<Arc<Transaction>>::max_allocation()+1) {
            smallest_disallowed_vec.push(Arc::new(tx.clone()));
        }
        let serialized = smallest_disallowed_vec.zcash_serialize_to_vec().expect("Serialization to vec must succeed");

        // Check that our smallest_disallowed_vec is only one item larger than the limit
        prop_assert!(((smallest_disallowed_vec.len() - 1) as u64) == <Arc<Transaction>>::max_allocation());
        // Check that our smallest_disallowed_vec is too big to be included in a valid block
        prop_assert!(serialized.len() as u64 > MAX_BLOCK_BYTES);
    }

    /// Verify the smallest disallowed vector of `Input`s is too large to fit in a Zcash block
    #[test]
    fn input_max_allocation_is_big_enough(input in Input::arbitrary()) {
```

File: zebra-chain/src/transaction/tests/preallocate.rs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        // Check that our smallest_disallowed_vec is only one item larger than the limit
+        prop_assert!(((smallest_disallowed_vec.len() - 1) as u64) == Input::max_allocation());
+        // Check that our smallest_disallowed_vec is too big to be included in a valid block
+        // Note that a serialized block always includes at least one byte for the number of transactions,
+        // so any serialized Vec<Input> at least MAX_BLOCK_BYTES long is too large to fit in a block.
+        prop_assert!(serialized.len() as u64 >= MAX_BLOCK_BYTES);
+    }
+    #[test]
+    /// Verify the smallest disallowed vector of `Output`s is too large to fit in a Zcash block
+    fn output_max_allocation_is_big_enough(output in Output::arbitrary()) {
```
Comment (teor2345) [inline zebra-chain/src/transaction/tests/preallocate.rs]:
```suggestion
    }

    /// Verify the smallest disallowed vector of `Output`s is too large to fit in a Zcash block
    #[test]
    fn output_max_allocation_is_big_enough(output in Output::arbitrary()) {
```

File: zebra-network/src/meta_addr/tests/preallocate.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    let serialized = addr
+        .zcash_serialize_to_vec()
+        .expect("Serialization to vec must succeed");
+    assert!(serialized.len() == META_ADDR_SIZE)
+}
+#[test]
+/// Verifies that...
+/// 1. The smallest disallowed vector of `MetaAddrs`s is too large to fit in a legal Zcash message
+/// 2. The largest allowed vector is small enough to fit in a legal Zcash message
+fn meta_addr_max_allocation_is_correct() {
```
Comment (teor2345) [inline zebra-network/src/meta_addr/tests/preallocate.rs]:
```suggestion
/// Confirm that each MetaAddr takes exactly META_ADDR_SIZE bytes when serialized.
/// This verifies that our calculated `TrustedPreallocate::max_allocation()` is indeed an upper bound.
#[test]
fn meta_addr_size_is_correct() {
    let addr = MetaAddr {
        addr: ([192, 168, 0, 0], 8333).into(),
        services: PeerServices::default(),
        last_seen: Utc.timestamp(1_573_680_222, 0),
        last_connection_state: PeerAddrState::Responded,
    };
    let serialized = addr
        .zcash_serialize_to_vec()
        .expect("Serialization to vec must succeed");
    assert!(serialized.len() == META_ADDR_SIZE)
}

/// Verifies that...
/// 1. The smallest disallowed vector of `MetaAddrs`s is too large to fit in a legal Zcash message
/// 2. The largest allowed vector is small enough to fit in a legal Zcash message
#[test]
fn meta_addr_max_allocation_is_correct() {
```

File: zebra-network/src/protocol/external/tests/preallocate.rs (hunk 1 above)
Comment (teor2345) [inline zebra-network/src/protocol/external/tests/preallocate.rs]:
```suggestion
/// Confirm that each InventoryHash takes exactly INV_HASH_SIZE bytes when serialized.
/// This verifies that our calculated `TrustedPreallocate::max_allocation()` is indeed an upper bound.
#[test]
fn inv_hash_size_is_correct() {
    let block_hash = block::Hash([1u8; 32]);
    let tx_hash = transaction::Hash([1u8; 32]);
    let inv_block = InventoryHash::Block(block_hash);
    let serialized_inv_block = inv_block
        .zcash_serialize_to_vec()
        .expect("Serialization to vec must succeed");
    assert!(serialized_inv_block.len() == INV_HASH_SIZE);

    let inv_filtered_block = InventoryHash::FilteredBlock(block_hash);
    let serialized_inv_filtered = inv_filtered_block
        .zcash_serialize_to_vec()
        .expect("Serialization to vec must succeed");
    assert!(serialized_inv_filtered.len() == INV_HASH_SIZE);

    let inv_tx = InventoryHash::Tx(tx_hash);
    let serialized_inv_tx = inv_tx
        .zcash_serialize_to_vec()
        .expect("Serialization to vec must succeed");
    assert!(serialized_inv_tx.len() == INV_HASH_SIZE);

    let inv_err = InventoryHash::Error;
    let serializd_inv_err = inv_err
        .zcash_serialize_to_vec()
        .expect("Serialization to vec must succeed");
    assert!(serializd_inv_err.len() == INV_HASH_SIZE)
}

/// Verifies that...
/// 1. The smallest disallowed vector of `InventoryHash`s is too large to fit in a legal Zcash message
/// 2. The largest allowed vector is small enough to fit in a legal Zcash message
#[test]
fn inv_hash_max_allocation_is_correct() {
```

File: zebra-chain/src/block/tests/preallocate.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
+//! Tests for trusted preallocation during deserialization.
+
+use super::super::{
+    header::MIN_COUNTED_HEADER_LEN, CountedHeader, Hash, Header, BLOCK_HASH_SIZE,
+    MAX_PROTOCOL_MESSAGE_LEN,
+};
+
+use crate::serialization::{TrustedPreallocate, ZcashSerialize};
```
Comment (dconnolly) [inline zebra-chain/src/block/tests/preallocate.rs]: Smush these together?

```suggestion
use crate::{
   block::{
       header::MIN_COUNTED_HEADER_LEN, CountedHeader, Hash, Header, BLOCK_HASH_SIZE,
       MAX_PROTOCOL_MESSAGE_LEN,
    },
    serialization::{TrustedPreallocate, ZcashSerialize}
};
```

File: zebra-chain/src/block/tests/preallocate.rs [hunk 7]
```diff
@@ -0,0 +1,98 @@
+//! Tests for trusted preallocation during deserialization.
+
+use crate::{
+   block::{
+       header::MIN_COUNTED_HEADER_LEN, CountedHeader, Hash, Header, BLOCK_HASH_SIZE,
+       MAX_PROTOCOL_MESSAGE_LEN,
+    },
+    serialization::{TrustedPreallocate, ZcashSerialize}
+};
```
Comment (teor2345) [inline zebra-chain/src/block/tests/preallocate.rs]: Rustfmt
```suggestion
use crate::{
    block::{
        header::MIN_COUNTED_HEADER_LEN, CountedHeader, Hash, Header, BLOCK_HASH_SIZE,
        MAX_PROTOCOL_MESSAGE_LEN,
    },
    serialization::{TrustedPreallocate, ZcashSerialize},
};
```

**Comment by teor2345** [comment]
This is blocking security fixes, so I merged it.

File: zebra-network/src/meta_addr/tests/prop.rs [hunk 8]
```diff
@@ -0,0 +1,29 @@
+use super::super::*;
```
Comment (dconnolly) [inline zebra-network/src/meta_addr/tests/prop.rs:1]: Could be constrained

**Review by dconnolly — APPROVED** [review]

File: zebra-network/src/meta_addr/tests/prop.rs (hunk 8 above)
Comment (teor2345) [reply zebra-network/src/meta_addr/tests/prop.rs:1]: Already fixed in #1985

**Pull request merged**